polkadot-parachains/westmint/src/lib.rs

```diff
@@ -456,6 +461,40 @@ pub type CurrencyTransactor = CurrencyAdapter<
 	(),
 >;
 
+pub struct AsPrefixedGeneralIndexFromLocalOrRemote<
+	LocalPrefix,
+	RemotePrefix,
+	AssetId,
+	ConvertAssetId,
+>(PhantomData<(LocalPrefix, RemotePrefix, AssetId, ConvertAssetId)>);
+impl<
+		LocalPrefix: Get<MultiLocation>,
+		RemotePrefix: Get<MultiLocation>,
+		AssetId: Clone,
+		ConvertAssetId: Convert<u128, AssetId>,
+	> Convert<MultiLocation, AssetId>
+	for AsPrefixedGeneralIndexFromLocalOrRemote<LocalPrefix, RemotePrefix, AssetId, ConvertAssetId>
+{
+	fn convert_ref(id: impl Borrow<MultiLocation>) -> result::Result<AssetId, ()> {
+		let prefix = LocalPrefix::get();
+		let remote_prefix = RemotePrefix::get();
+		let id = id.borrow();
+		match id.interior().at(prefix.interior().len()) {
+			Some(Junction::GeneralIndex(id)) => ConvertAssetId::convert_ref(id),
+			_ => match id.interior().at(remote_prefix.interior().len()) {
+				Some(Junction::GeneralIndex(id)) => ConvertAssetId::convert_ref(id),
+				_ => Err(()),
+			},
+		}
+	}
+	fn reverse_ref(what: impl Borrow<AssetId>) -> result::Result<MultiLocation, ()> {
+		let mut location = LocalPrefix::get();
+		let id = ConvertAssetId::reverse_ref(what)?;
+		location.push_interior(Junction::GeneralIndex(id)).map_err(|_| ())?;
+		Ok(location)
+	}
+}
```

Comment:
Thank you for the contribution from my side as well, btw. :pray: 

Here are the adjustments I'd like to see:
+ [ ] Please move this into parachains-common. (Statemint and Statemine will have to use the same.)
+ [ ] This also should have a few tests:
  + matching the local prefix
  + matching the remote prefix
  + the failure cases
+ [ ] Please add docs
+ [ ] Also you skipped the prefix checking from the original here:
https://github.com/paritytech/polkadot/blob/55872676fae63399eb111a60ba27cb735cbaead6/xcm/xcm-builder/src/fungibles_adapter.rs#L41-L46
You should be able to use `match_and_split` instead, though: https://github.com/paritytech/polkadot/blob/cc24fb87b702db6510d08c689e442b6be384d798/xcm/src/v1/multilocation.rs#L249

+ We should probably also add (something like) this to the xcm-builder as @KiChjang suggests. (Would be great if you could open that PR as well to upstream this.)